### PR TITLE
Clickable variable names 98

### PR DESCRIPTION
**Pull request by sandcha**
Connected to #98 

In a variable page, replace any reference to an OpenFisca variable by a link to its page.

### CODE REVIEW

File: src/components/variable.jsx [hunk 1]
```diff
@@ -107,14 +135,29 @@ const Variable = React.createClass({
...
+  link(substring, quote) {
+    if (!substring.includes(" ") && this.props.variables[substring]) {
+      return <Link key={substring} to={substring} onClick={this.navigateProgramatically}>{quote + substring + quote}</Link>
+    } 
+    return quote + substring + quote
+  },
+  // Change every OpenFisca variable in the formula by a link to the variable page:
+  renderLinkedFormulaVariables(formula) {
+    return formula.split("'").map((substring, index) => {
+      return (index % 2 != 0) ?
```
Comment (fpagnoux) [inline src/components/variable.jsx]: What if there is a single `'` in the comments of the formula ?

For instance:
```py
def formula(famille, period):
    # This ' will break the algorithm
    x = famille('x', period)
    y = famille('y', period)
    return x + y
```

File: src/server/webpack-dev-server.js [hunk 2]
```diff
@@ -14,6 +14,7 @@ new WebpackDevServer(webpack(config), {
   noInfo: true,
   publicPath: config.output.publicPath,
   stats: { colors: true },
+  quiet: true,
```
Comment (fpagnoux) [inline src/server/webpack-dev-server.js]: > 🎉

File: src/components/variable.jsx [hunk 3]
```diff
@@ -107,14 +135,29 @@ const Variable = React.createClass({
...
+                <pre><code>{this.renderLinkedFormulaVariables(formulas[date].content)}</code></pre>
               </div>
             )
           })
         }
       </div>
     )
   },
+  link(substring, quote) {
+    if (!substring.includes(" ") && this.props.variables[substring]) {
```
Comment (fpagnoux) [inline src/components/variable.jsx]: > !substring.includes(" ")

Seems like a fair optimization: you only keep single words 👍 . Maybe a comment to explain why ?

Comment (fpagnoux) [inline src/components/variable.jsx]: According to the functions naming, I'd expect `renderLinkedFormulaVariables` to check that the substrings are valid variables, and `link` to just build a `Link`.

File: src/components/variable.jsx [hunk 4]
```diff
@@ -20,6 +25,29 @@ const Variable = React.createClass({
   getTodayInstant() {
     return new Date().toJSON().slice(0, 10)
   },
+  navigateProgramatically(e) {
```
Comment (fpagnoux) [inline src/components/variable.jsx]: This seems complex. It seems that we are re-implementing routing. 

What problem that you encountered brought you to write this ? Did you get help from anyone who knows react well ?

**Comment by fpagnoux** [comment]
Also, an integration test would be welcome :).

File: src/components/variable.jsx (hunk 4 above)
Comment (fpagnoux) [inline src/components/variable.jsx]: Maybe check there: https://github.com/ReactTraining/react-router/issues/4975

Comment (fpagnoux) [inline src/components/variable.jsx]: No, this seems to be the problem: https://stackoverflow.com/questions/32261441/component-does-not-remount-when-route-parameters-change

File: src/components/variable.jsx [hunk 5]
```diff
@@ -107,14 +135,29 @@ const Variable = React.createClass({
...
   },
+  link(substring, quote) {
+    if (!substring.includes(" ") && this.props.variables[substring]) {
+      return <Link key={substring} to={substring} onClick={this.navigateProgramatically}>{quote + substring + quote}</Link>
+    } 
+    return quote + substring + quote
+  },
+  // Change every OpenFisca variable in the formula by a link to the variable page:
+  renderLinkedFormulaVariables(formula) {
+    return formula.split("'").map((substring, index) => {
```
Comment (sandcha) [inline src/components/variable.jsx]: Ok but we might have to loop on the splits after the first map.

File: src/components/variable.jsx [hunk 6]
```diff
@@ -107,14 +108,44 @@ const Variable = React.createClass({
...
+    })
+  },
+  //Change every OpenFisca variable in the formula by a link to the variable page:
+  renderLinkedFormulaVariables(formula) {
+    //Split on double quotes first (preventing collision with Link):
+    var splits = this.linkCodeSplits(formula.split('"'), '"') 
+    for (var s of splits) {
+      s = this.linkCodeSplits(s.split("'"), "'")
+    }
+    return s
```
Comment (fpagnoux) [inline src/components/variable.jsx]: That seems to be only returning the last split... I doubt this works in a variable with several variables calls surrounded by `""`

File: src/components/variable.jsx [hunk 7]
```diff
@@ -107,14 +108,44 @@ const Variable = React.createClass({
...
+    return (!substring.includes(" ") && this.props.variables[substring])
+  },
+  linkCodeSplits(splits, separator) {
+    var previousIsLink = false
+    return splits.map((substring, index) => {
+      if (this.isVariable(substring)) {
+        substring = this.link(substring)
+        previousIsLink = true
+      } else {
+        substring = previousIsLink ? separator + substring : substring  //No jsx and separator concatenation.
```
Comment (fpagnoux) [inline src/components/variable.jsx]: I don't really understand... Why do we need to add a separator only if the previous item was a `Link` ?

**Pull request merged**